fontforge/python.c

```diff
@@ -14066,6 +14066,150 @@ return (NULL);
+    struct otfname *on;
+
+    if ( CheckIfFontClosed(self) )
+        return( -1 );
+
+    if ( !PyTuple_Check(value) ) {
+        PyErr_Format(PyExc_TypeError, "Style set names must be set as a tuple.");
+        return( -1 );
+    }
+    rows = PyTuple_Size(value);
+    for ( i=0; i<rows; ++i ) {
+        ss_names_tuple = PyTuple_GetItem(value, i);
+        if ( !PyTuple_Check(ss_names_tuple) ) {
+            PyErr_Format(PyExc_TypeError, "Style set name specification must be a tuple.");
+            return( -1 );
+        }
+        if ( PyTuple_Size(ss_names_tuple)!=3 ) {
+            PyErr_Format(PyExc_TypeError, "Style set name specification must have 3 elements (language, style set and name).");
+            return( -1 );
+        }
+
+        PyObject *lang_py = PyTuple_GetItem(ss_names_tuple, 0);
+        if ( PyUnicode_Check(lang_py) ) {
+            if ( FindFlagByName(sfnt_name_mslangs, PyUnicode_AsUTF8(lang_py))==FLAG_UNKNOWN  ) {
+                PyErr_Format(PyExc_TypeError, "The first element of a style set name specification must be a valid language name if it is a string.");
+                return( -1 );
+            }
+        } else if ( PyLong_Check(lang_py) ) {
+            if ( strcmp(NOUI_MSLangString(PyLong_AsLong(lang_py)), _("Unknown"))==0 ) {
+                PyErr_Format(PyExc_TypeError, "The first element of a style set name specification must be a valid language id if it is an integer.");
+                return( -1 );
+            }
+        } else {
+            PyErr_Format(PyExc_TypeError, "The first element of a style set name specification must be a string or an integer.");
+            return( -1 );
+        }
+
+        PyObject *tag_str_py = PyTuple_GetItem(ss_names_tuple, 1);
+        if ( !PyUnicode_Check(tag_str_py) ) {
+            PyErr_Format(PyExc_TypeError, "The second element of a style set name specification must be a string.");
+            return( -1 );
+        }
+        tag_str = copy(PyUnicode_AsUTF8(tag_str_py));
+        is_valid = 0;
+        if ( strlen(tag_str)== 4 && tag_str[0]=='s' && tag_str[1]=='s'
+            && '0'<=tag_str[2] && tag_str[2]<='9' && '0'<=tag_str[3] && tag_str[3]<='9'
+        ) {
+            int ss_num = 10*(tag_str[2]-'0') + tag_str[3]-'0';
+            is_valid = 1 <= ss_num && ss_num <= 20;
+        }
+        if ( !is_valid ) {
+            PyErr_Format(PyExc_TypeError, "The second element of a style set name specification must be a valid style set tag.");
+            return( -1 );
+        }
+
+        if ( !PyUnicode_Check(PyTuple_GetItem(ss_names_tuple, 2)) ) {
+            PyErr_Format(PyExc_TypeError, "The third element of a style set name specification must be a string.");
+            return( -1 );
+        }
+    }
+
+    /* check for duplicates */
+    for ( i=0; i<rows; ++i ) {
+        for ( j=i+1; j<rows; ++j ) {
+            if ( PyTuple_GetItem(PyTuple_GetItem(value, i), 0) == PyTuple_GetItem(PyTuple_GetItem(value, j), 0) &&
```

Comment:
I'd be surprised (but not flabbergasted) if these comparisons work. They're pointer comparisons, and if they work it implies that Python is doing a very good job making sure that tuples are unique internally. 

Might `PyObject_RichCompareBool` with `PyEQ` be what we want here?